### ISSUE TITLE
Possiblity of off-center camera?

### ISSUE THREAD

**Issue by ieremius** [graphdeco-inria/gaussian-splatting#650]
Regarding

__device__ float3 computeCov2D(const float3& mean, float focal_x, float focal_y, float tan_fovx, float tan_fovy, const float* cov3D, const float* viewmatrix)
{
	// The following models the steps outlined by equations 29
	// and 31 in "EWA Splatting" (Zwicker et al., 2002). 
...

Just before those formulas in Zwicker et al., 2002 they write, "Camera space is defined such that the origin of the camera coordinate system is at the center of projection."

And obviously computeCov2D() has fov parameters.

So this is all corresponds to glm::perspective() - a centered camera.

Has anyone considered what it would take to support a case like glm::frustum(), where the camera is off-center? Could there be an alternate version of eqution 29 to support this? Or is the problem deeper than that? Should I just forget about it right now?

I'm still trying to get up to speed with the existing math from Zwicker, but wondered if someone could point me in a direction.

I'm referening here only to rasterization in a viewer, (hopefully) not training.

**Comment by Ryoto-Kato** [comment]
I think this will solve your problem.
https://github.com/graphdeco-inria/gaussian-splatting/issues/144#issuecomment-1938504456

**Comment by rocksat** [comment]
I have the same question, and just by modify the `getProjectionMatrix` didn't solved my problem, looks like it's need to change the CUDA differentiable rasterization part.

**Comment by ieremius** [comment]
@rocksat, yes exactly. I've tried modifying the projection matrix as well, to represent the viewer moving around in the x and y directions in the frustum...what I then see is just horizontal and vertical translation, but not the perspective distortion I'm looking for. Which makes sense, since the CUDA rasterization part appears to make further assumptions about the viewer being in the center of the frustum.
